server/src/controllers/user.ts

```diff
@@ -3,99 +3,108 @@
 // SPDX-License-Identifier: MIT
 
 import { Request, Response } from 'express';
-import { Course } from './course';
+import models from '../database/models';
+import Course from '../database/models/course';
+import CourseInstance from '../database/models/courseInstance';
+import CourseTranslation from '../database/models/courseTranslation';
+import { CourseData, Language } from './course';
 
 export interface TeacherCourses {
-  current: Array<Course>,
-  previous: Array<Course>
+  current: Array<CourseData>,
+  previous: Array<CourseData>
 }
 
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
-async function getTeacherCourses(userId: number): Promise<TeacherCourses> {
-  return {
-    current: [
-      {
-        id: 0,
-        courseCode: 'CS-E4580',
-        minCredits: 5,
-        maxCredits: 5,
-        department: {
-          fi: '',
-          sv: '',
-          en: 'Department of Computer Science'
-        },
-        name: {
-          fi: '',
-          sv: '',
-          en: 'Programming Parallel Computers D'
-        },
-        evaluationInformation: {
-          fi: '',
-          sv: '',
-          en: ''
-        }
+export async function getUserCourses(req: Request, res: Response): Promise<void> {
+  try {
+    const teacherCourses: TeacherCourses = { current: [], previous: [] };
+
+    // TODO: Go through course_role instead
+    let instances: Array<CourseInstance> = await models.CourseInstance.findAll({
```

Comment:
We could query all courses belonging to the teacher here, and avoid querying inside the loop later on.
For an example with:

```typescript
let instances: Array<Course> = await models.Course.findAll({
  attributes: ['id', 'courseCode', 'minCredits', 'maxCredits'],
  include: [{
    model: CourseInstance,
    where: {
      responsibleTeacher: Number(req.params.userId)
    },
    order: ['startDate', 'DESC']
  },
  {
    model: CourseTranslation,
    attributes: ['language', 'courseName', 'department'],
  }],
});
```

Should fetch all courses, translations and their instances (arranged by start date descending) with teacher id. Then it would be just a matter of reformatting the data to match the model we want to return to the user.

Reply:
Yes, that is much better. It's been a while since I've used SQL (and Sequelize never) so I didn't remember/know how to do this in a better way, thanks!